### PR TITLE
Add option to use DeepGemm contiguous grouped gemm kernel for fused MoE operations.

### PR DESCRIPTION
**Pull request by bnellnm**
Add the option to use DeepGemm's grouped gemm kernel for `fused_moe` operations.

### Benchmark setup
```
python3 benchmarks/kernels/benchmark_moe.py --dtype=fp8_w8a8 -tp ${TP} --model deepseek-ai/DeepSeek-V3 --trust-remote-code --use-deep-gemm
Note: deepseek-ai/DeepSeek-R1 has similar problem sizes
E=256
N=4096 @ TP=1, 2048 @ TP=2, 1024 @ TP=4, 512 @ TP=8
K=7168
topk=8
```
### H100 results
|Batch | DeepGemm TP=1| Triton TP=1| DeepGemm TP=2| Triton TP=2| DeepGemm TP=4| Triton TP=4| DeepGemm TP=8| Triton TP=8|
|------|-------------|-------|-------------|-------|-------------|-------|-------------|-------|
|128 | 6355.22 us | 12480.49 us | 3745.24 us | 6025.39 us | 2529.50 us | 3121.16 us | 1936.17 us | 1732.33 us|
|256 | 6599.51 us | 12782.98 us | 3917.15 us | 6164.44 us | 2669.84 us | 3236.88 us | 2051.67 us | 1728.73 us|
|512 | 7017.80 us | 13017.54 us | 4270.21 us | 6631.64 us | 3000.99 us | 3433.29 us | 2331.06 us | 1810.36 us|
|1024 | 7850.04 us | 13322.31 us | 4874.43 us | 6800.41 us | 3575.18 us | 3546.83 us | 2886.08 us | 1892.75 us|
|1536 | 8691.74 us | 13777.98 us | 5559.64 us | 7037.34 us | 4111.26 us | 3677.90 us | 3423.49 us | 1988.70 us|
|2048 | 9545.87 us | 19839.38 us | 6217.40 us | 10111.40 us | 4699.03 us | 5230.76 us | 3974.65 us | 2818.65 us|
|3072 | 11252.75 us | 27189.87 us | 7609.45 us | 13845.57 us | 5916.31 us | 7172.64 us | 5093.12 us | 3858.15 us|
|4096 | 12843.01 us | 33834.67 us | 8669.07 us | 17239.74 us | 7167.47 us | 8927.87 us | 6225.71 us | 4789.98 us|

### H200 results 
|Batch | DeepGemm TP=1| Triton TP=1| DeepGemm TP=2| Triton TP=2| DeepGemm TP=4| Triton TP=4| DeepGemm TP=8| Triton TP=8|
|------|-------------|-------|-------------|-------|-------------|-------|-------------|-------|
|128 | 5206.92 us | 10821.89 us | 3205.37 us | 5596.44 us | 2257.24 us | 2710.88 us | 1824.80 us | 1472.32 us|
|256 | 5415.36 us | 11069.98 us | 3373.36 us | 5587.06 us | 2394.34 us | 2785.30 us | 1952.46 us | 1528.53 us|
|512 | 5825.33 us | 13446.07 us | 3711.86 us | 6843.68 us | 2732.21 us | 3568.38 us | 2229.71 us | 1876.63 us|
|1024 | 6599.35 us | 13754.83 us | 4362.23 us | 7016.18 us | 3287.25 us | 3681.09 us | 2754.67 us | 1946.25 us|
|1536 | 7387.39 us | 14218.68 us | 5024.80 us | 7249.93 us | 3851.06 us | 3812.64 us | 3308.97 us | 2049.49 us|
|2048 | 8196.78 us | 20468.26 us | 5683.06 us | 10424.77 us | 4432.50 us | 5384.37 us | 3836.07 us | 2898.91 us|
|3072 | 9842.72 us | 28046.94 us | 7005.50 us | 14262.27 us | 5623.91 us | 7397.70 us | 4915.28 us | 3985.38 us|
|4096 | 11375.09 us | 34898.63 us | 8171.68 us | 17769.66 us | 6829.79 us | 9232.80 us | 6017.09 us | 4942.76 us|

I've run server benchmarks using DeepSeek-R1 but right now the Triton and DeepGemm kernels are at parity due to the TP=8 performance of the DeepGemm kernels.  I think this can be improved by permute/unpermute ops (similar to #14568) and/or using EP instead of TP for fused_moe ops.

Note: tested with DeepGemm https://github.com/deepseek-ai/DeepGEMM/commit/e1c070fbef0d522fda469f2720c4402969afd5ac.  Some newer revisions result in wrong answers.

cc @mgoin, @ElizaWszola, @LucasWilkinson

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 1]
```diff
@@ -1277,21 +1351,114 @@ def fused_experts(hidden_states: torch.Tensor,
...
+            block_shape, allow_deep_gemm)
+
+
+def _fp8_perm(m: torch.Tensor, idx: torch.Tensor) -> torch.Tensor:
+    """
+    A permutation routine that works on fp8 types.
+    """
+    if torch.is_floating_point(m) and torch.finfo(m.dtype).bits == 8:
+        return m.view(dtype=torch.uint8)[idx,
+                                         ...].view(dtype=torch.float8_e4m3fn)
```
Comment (youkaichao) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1378]: for the second view, do we need to use `m.dtype` instead of `torch.float8_e4m3fn`?

**Review by youkaichao — COMMENTED** [review]
how large is the wheel? do we want to ship it by default?

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 1 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1378]: Good catch. Thanks!

**Comment by bnellnm** [comment]
> how large is the wheel? do we want to ship it by default?

The DeepGemm repo is around ~200Mb. Most of that comes from the cutlass files.  I don't think there are any whl files for this repo though.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 2]
```diff
@@ -1277,21 +1351,114 @@ def fused_experts(hidden_states: torch.Tensor,
...
+    sorted_token_ids, expert_ids, num_tokens_post_padded = (
+        moe_align_block_size(curr_topk_ids, block_m, global_num_experts,
+                             expert_map))
+
+    inv_perm: Optional[torch.Tensor] = None
+
+    if use_dg:
+        # Replicate activations
+        max_token_id = top_k_num * tokens_in_chunk
+        pad_size = (((sorted_token_ids.numel() + block_m - 1) // block_m) *
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1406]: nit: in `utils.py` we have `round_up`, i.e. this could be `round_up(sorted_token_ids.numel(), block_m) - - sorted_token_ids.numel()`

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 3]
```diff
@@ -1277,21 +1351,114 @@ def fused_experts(hidden_states: torch.Tensor,
...
+        expert_ids = torch.repeat_interleave(expert_ids, block_m, dim=0)
+        inv_perm = torch.argsort(sorted_token_ids)[:max_token_id]
+
+        curr_hidden_states = curr_hidden_states.view(
+            curr_hidden_states.shape[0], -1,
+            curr_hidden_states.shape[1]).repeat(1, top_k_num, 1).reshape(
+                -1, curr_hidden_states.shape[1])
+
+        # Permute according to sorted token ids.
+        curr_hidden_states = _fp8_perm(curr_hidden_states, sorted_token_ids)
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1424]: instead of all of this:
```
        curr_hidden_states = curr_hidden_states.view(
             curr_hidden_states.shape[0], -1,
             curr_hidden_states.shape[1]).repeat(1, top_k_num, 1).reshape(
                 -1, curr_hidden_states.shape[1])

         # Permute according to sorted token ids.
         curr_hidden_states = _fp8_perm(curr_hidden_states, sorted_token_ids)
```
can we do:
```
         # Permute according to sorted token ids.
         curr_hidden_states = _fp8_perm(curr_hidden_states, sorted_token_ids // top_k)
```
?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1424]: Yes, thanks!  That's a good observation.

File: vllm/model_executor/layers/quantization/fp8.py [hunk 4]
```diff
@@ -409,6 +424,7 @@ class Fp8MoEMethod(FusedMoEMethodBase):
     def __init__(self, quant_config: Fp8Config):
         self.quant_config = quant_config
         self.block_quant = self.quant_config.weight_block_size is not None
+        self.allow_deep_gemm = use_deep_gemm
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/fp8.py:427]: nit: for consistency it might be nice to use `allow_deep_gemm` or `use_deep_gemm` everywhere, I find the mix a bit confusing (although I do understand it now so not critical)

File: vllm/model_executor/layers/quantization/fp8.py [hunk 5]
```diff
@@ -566,6 +582,17 @@ def process_weights_after_loading(self, layer: Module) -> None:
                                                       requires_grad=False)
                 layer.w2_weight = torch.nn.Parameter(shuffled_w2,
                                                      requires_grad=False)
+
+            # DeepGemm scales need to be transposed and aligned.  We try to do
+            # it ahead of time for performance reasons.
+            if self.allow_deep_gemm:
+                if _is_col_major(layer.w13_weight_scale_inv):
+                    layer.w13_weight_scale_inv = \
+                        dg.get_col_major_tma_aligned_tensor(layer.w13_weight_scale_inv).contiguous()
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/fp8.py:591]: Does `get_col_major_tma_aligned_tensor` require that its column major? for a cursory reading of https://github.com/deepseek-ai/DeepGEMM/blob/c57699ac933a93651c34d365797c2d8b41a4765b/deep_gemm/jit_kernels/utils.py#L77-L109 it seems like it could handle none col-major layouts? If does should we assert if its not col-major and we don't call `get_col_major_tma_aligned_tensor`?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 6]
```diff
@@ -203,6 +203,8 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         layer.w13_weight_scale = torch.nn.Parameter(max_w13_scales,
                                                     requires_grad=False)
 
+        # TODO: do we need to do deep gemm alignment here?
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:206]: nit: is this comment needed? it doesnt appear that compressed tensors currently supports DeepGEMM

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 7]
```diff
@@ -1337,7 +1412,91 @@ def fused_experts(hidden_states: torch.Tensor,
...
+                              inv_perm: Optional[torch.Tensor],
+                              m_indices: torch.Tensor, topk: int,
+                              num_groups: int, K: int,
+                              topk_weight: torch.Tensor,
+                              topk_ids: torch.Tensor, use_dg: bool) -> None:
+    """
+    If needed (for DeepGemm), unpermute the final result and apply topk_weights.
+    Then perform the final reduction on the hidden states.
+    """
+    if use_dg:
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1485]: potential future pr: to avoid all the `use_dg` calls it might be nice to: https://github.com/vllm-project/vllm/pull/13972#discussion_r2010289695

**Review by LucasWilkinson — COMMENTED** [review]
Overall looks pretty good! left a few more comments

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 8]
```diff
@@ -1337,7 +1412,91 @@ def fused_experts(hidden_states: torch.Tensor,
...
+    tokens_in_chunk, _ = curr_hidden_states.shape
+
+    sorted_token_ids, expert_ids, num_tokens_post_padded = (
+        moe_align_block_size(curr_topk_ids, block_m, global_num_experts,
+                             expert_map))
+
+    inv_perm: Optional[torch.Tensor] = None
+
+    if use_dg:
+        # Replicate activations
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1450]: It's not.  I'll remove it.

File: vllm/model_executor/layers/quantization/fp8.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/quantization/fp8.py:591]: I did this to try to avoid the transpose at runtime.  I check for col major so that we don't accidentally transpose the weights but fail to run the kernel using DeepGemm for reasons that can't be determine this far ahead of time.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 6 above)
Comment (bnellnm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:206]: I wasn't sure whether compressed tensors could use DeepGemm or not.  I'll remove the comment,

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 7 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1485]: I agree the flags are not nice.  Pulling the DeepGemm implementation out into a separate function would end up being just as messy since I require the Triton implementation for a fallback when `M` is not a multiple of 128.

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1485]: I'll see if I can think up a nicer way to do this.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 9]
```diff
@@ -1428,7 +1575,56 @@ def fused_experts_impl(hidden_states: torch.Tensor,
...
+
+    cache1_view: Tuple[int, ...] = ()
+    cache2_view: Tuple[int, ...] = ()
+    cache3_view: Tuple[int, ...] = ()
+
+    if use_dg:
+        # If M is not divisible by the block size we run the largest
+        # chunk we can using DeepGemm, the remainder is handed off to
+        # the Triton kernels.
+        if M % block_m != 0:
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:1589]: im a bit confused on why this is required, doesn't `moe_align_block_size` pad so we are block aligned after permuting?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:1589]: `moe_align_block_size` doesn't literally pad the `M` dimension it comes up with `num_tokens_post_padded` which the Triton kernels use to skip virtual padding.  This isn't useful for DeepGemm unless we want to actually physically pad the activations.

Comment (LucasWilkinson) [reply vllm/model_executor/layers/fused_moe/fused_moe.py]: For posterity, @bnellnm spoke offline, after the perm the activations have shape `sorted_token_ids.shape(0)` so 
`moe_align_block_size` does cause the activation to get padded meaning this check can be dropped :+1:

https://github.com/vllm-project/vllm/pull/13932/commits/160fde46eabb2f8950bb346aaf093d89c468b1ca

**Review by LucasWilkinson — APPROVED** [review]
LGTM now, thanks for the updates!

**Comment by youkaichao** [comment]
@bnellnm need to solve merge conflicts.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 10]
```diff
@@ -718,7 +778,16 @@ def invoke_fused_moe_kernel(A: torch.Tensor,
     grid = lambda META: (triton.cdiv(EM, META['BLOCK_SIZE_M']) * triton.cdiv(
         B.shape[1], META['BLOCK_SIZE_N']), )
 
-    if (use_int8_w8a16 or use_int4_w4a16) and \
+    if use_dg:
+        assert use_fp8_w8a8
+        # Note: we never apply the topk_weights here since it requires
+        # unpermuting and resizing the output.  This goes against the
+        # existing interface as the `mul_routed_weight` argument is
+        # ignored.  The weights are applied in _moe_unpermute.
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:786]: should this be `The weights are applied in _moe_unpermute_and_reduce`?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:786]: Yeah, I forgot to update the comment when I renamed the function.

**Review by tlrmchlsmth — COMMENTED** [review]
The `expert_map` doesn't work if it has `-1` in it, right? (so basically it doesn't work at all right now)

Do we have an assert anywhere for this?

**Comment by bnellnm** [comment]
> The `expert_map` doesn't work if it has `-1` in it, right? (so basically it doesn't work at all right now)
> 
> Do we have an assert anywhere for this?

I've added an assert and a check in _valid_deep_gemm.

File: vllm/model_executor/layers/quantization/fp8.py [hunk 11]
```diff
@@ -37,6 +37,21 @@
...
 logger = init_logger(__name__)
 
+allow_deep_gemm = False
+if envs.VLLM_USE_DEEP_GEMM:
+    try:
+        import deep_gemm as dg
+        logger.info("Using DeepGemm for fused MoE.")
+        allow_deep_gemm = True
+    except ImportError:
+        logger.warning("Failed to import DeepGemm kernels.")
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/fp8.py:47]: It would be nice to additionally check:
```
current_platform.is_cuda() and current_platform.has_device_capability(90)
````
and return false and log a warning otherwise

**Review by tlrmchlsmth — APPROVED** [review]
LGTM

**Review by tlrmchlsmth — APPROVED** [review]
